### DISCUSSION TITLE
change  some parameters simultaneously

### DISCUSSION THREAD

**Discussion by alirahdarian** [pybamm-team/PyBaMM#4042] [Q&A]
I use the following code to change 5 parameters (5 inputs) in the model but it returns an error while it works when I change only two parameters. I want to know how can I change these 5 parameters simultaneously.

import liionpack as lp
import pybamm
import numpy as np
I_mag = 5.0
OCV_init = 4.0 # used for intial guess
Ri_init = 5e-2 # used for intial guess
R_busbar = 1.5e-3
R_connection = 1e-2
Np = 2
Ns = 1
Nbatt = Np * Ns
netlist = lp.setup_circuit(
Np=Np, Ns=Ns, Rb=R_busbar, Rc=R_connection, Ri=Ri_init, V=OCV_init, I=I_mag
)
#print(netlist)

experiment = pybamm.Experiment(
[
"Charge at 2 A for 30 minutes",
"Rest for 15 minutes",
"Discharge at 2 A for 30 minutes",
"Rest for 15 minutes",
],
period="10 seconds",
)
parameter_values = pybamm.ParameterValues("Chen2020")

parameter_values.update({"Negative electrode diffusivity [m2.s-1]": "[input]"})
parameter_values.update({"Positive electrode diffusivity [m2.s-1]": "[input]"})
parameter_values.update({"Maximum concentration in negative electrode [mol.m-3]": "[input]"})
parameter_values.update({"Maximum concentration in positive electrode [mol.m-3]": "[input]"})
parameter_values.update({"SEI resistivity [Ohm.m]": "[input]"})

inputs = {"Negative electrode diffusivity [m2.s-1]": [3.3e-14,3.3e-14],
"Positive electrode diffusivity [m2.s-1]": [4e-15,4e-15],
"Maximum concentration in negative electrode [mol.m-3]": [33133.0,33133.0],
"Maximum concentration in positive electrode [mol.m-3]": [63104.0,63104.0],
"SEI resistivity [Ohm.m]": [200000.0,200000]
}

SPMe = pybamm.models.full_battery_models.lithium_ion.SPMe()
output = lp.solve(
netlist=netlist,
parameter_values=parameter_values,
experiment=experiment,
inputs=inputs,
initial_soc=0.5
)

print(parameter_values)

#lp.plot_output(output)
#lp.show_plots()

**Comment by rtimms** [comment]
Can you post the error?

**Reply by alirahdarian** [reply]
this is the error:

Traceback (most recent call last):
  File "/home/ali/New_folder/pack/agedcell.py", line 43, in <module>
    output = lp.solve(
  File "/home/ali/.local/lib/python3.10/site-packages/liionpack/solver_utils.py", line 427, in solve
    output = rm.solve(
  File "/home/ali/.local/lib/python3.10/site-packages/liionpack/solvers.py", line 201, in solve
    self.setup_actors(nproc, self.inputs_dict, initial_soc)
  File "/home/ali/.local/lib/python3.10/site-packages/liionpack/solvers.py", line 471, in setup_actors
    a.setup(
  File "/home/ali/.local/lib/python3.10/site-packages/liionpack/solvers.py", line 46, in setup
    _, _ = lp.update_init_conc(parameter_values, initial_soc, update=True)
  File "/home/ali/.local/lib/python3.10/site-packages/liionpack/sim_utils.py", line 53, in update_init_conc
    x, y = lp.get_initial_stoichiometries(SoC, param)
  File "/home/ali/.local/lib/python3.10/site-packages/liionpack/sim_utils.py", line 31, in get_initial_stoichiometries
    return esoh_solver.get_initial_stoichiometries(initial_soc)
  File "/home/ali/.local/lib/python3.10/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py", line 691, in get_initial_stoichiometries
    x_0, x_100, y_100, y_0 = self.get_min_max_stoichiometries()
  File "/home/ali/.local/lib/python3.10/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py", line 759, in get_min_max_stoichiometries
    Q_n = parameter_values.evaluate(param.n.Q_init)
  File "/home/ali/.local/lib/python3.10/site-packages/pybamm/parameters/parameter_values.py", line 774, in evaluate
    raise ValueError("symbol must evaluate to a constant scalar or array")
ValueError: symbol must evaluate to a constant scalar or array

**Comment by NicolaCourtier** [comment]
Hi @rtimms, I came across this error and believe the cause is the assumption that neither maximum concentration is an input in `pybamm.ParameterValues.set_initial_stoichiometries`, [line 347](https://github.com/pybamm-team/PyBaMM/blob/v24.11.2/src/pybamm/parameters/parameter_values.py#L347).

I think this can be rectified by taking the value from `inputs` if given, instead of evaluating the entry in `param`.

See also `set_initial_stoichiometries_half_cell`, [line 307](https://github.com/pybamm-team/PyBaMM/blob/v24.11.2/src/pybamm/parameters/parameter_values.py#L347).

**Reply by rtimms** [reply]
Thanks! You can actually pass `inputs` as a kwarg to `evaluate` which is what we should do here and requires no additional logic

**Reply by JulWilh** [reply]
I have kind of the same error: I update the Prada2013 values with a user defined function for Entropic Change (Pybamm interpolant).
My parameter file named "myFile" contains: 

```
def Negative_Entropy(sto):
    xval=np.array( [....])
    facB=np.array( [....])
    du_val =  pybamm.Interpolant(xval,facB,sto,interpolator="linear")
    
    return du_val
    
def get_parameter_values():# Call dict via a function to avoid errors when editing in place
    return { ...
    "Negative electrode OCP entropic change [V.K-1]": Negative_Entropy,
    ...
    }
```
    

In my code I import the new parameter values: 

```
import myFile as inputParams
paramset = inputParams.get_parameter_values()   # paramset is used to create pybamm.ParameterValues

# Update parameter with above stated set
parameter_values = pybamm.ParameterValues("Prada2013")
parameter_values.update(paramset, check_already_exists=False)
```

When running a simulation it delivers the following error: 

Traceback (most recent call last):

 File myPath/lib64/python3.11/site-packages/pybamm/parameters/parameter_values.py:867 in evaluate
    return processed_symbol.evaluate(inputs=inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/expression_tree/binary_operators.py:163 in evaluate
    left = self.left.evaluate(t, y, y_dot, inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/expression_tree/binary_operators.py:164 in evaluate
    right = self.right.evaluate(t, y, y_dot, inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/expression_tree/binary_operators.py:164 in evaluate
    right = self.right.evaluate(t, y, y_dot, inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/expression_tree/functions.py:131 in evaluate
    evaluated_children = [

  File myPath/lib64/python3.11/site-packages/pybamm/expression_tree/functions.py:132 in <listcomp>
    child.evaluate(t, y, y_dot, inputs) for child in self.children

  File myPath/lib64/python3.11/site-packages/pybamm/expression_tree/symbol.py:830 in evaluate
    return self._base_evaluate(t, y, y_dot, inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/expression_tree/independent_variable.py:105 in _base_evaluate
    raise ValueError("t must be provided")

ValueError: t must be provided


The above exception was the direct cause of the following exception:
```

Traceback (most recent call last):

  File myPath/lib64/python3.11/site-packages/spyder_kernels/customize/utils.py:209 in exec_encapsulate_locals
    exec_fun(compile(code_ast, filename, "exec"), globals)

  File myPath/RunSingle.py:166
    _,_,firstnest,parameter_values_first = simulate_voltage_with_params(actual_nest, drive_cycle,parameter_values)

  File myPath/RunSingle.py:122 in simulate_voltage_with_params
    sim.solve([0, drive_cycle[-1, 0]],initial_soc=initialSOC)

  File myPath/lib64/python3.11/site-packages/pybamm/simulation.py:478 in solve
    self.build(initial_soc=initial_soc, inputs=inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/simulation.py:325 in build
    self.set_initial_soc(initial_soc, inputs=inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/simulation.py:295 in set_initial_soc
    self._unprocessed_parameter_values.set_initial_stoichiometries(

  File myPath/lib64/python3.11/site-packages/pybamm/parameters/parameter_values.py:375 in set_initial_stoichiometries
    x, y = pybamm.lithium_ion.get_initial_stoichiometries(

  File myPath/lib64/python3.11/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:923 in get_initial_stoichiometries
    return esoh_solver.get_initial_stoichiometries(initial_value, tol, inputs=inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:696 in get_initial_stoichiometries
    x_0, x_100, y_100, y_0 = self.get_min_max_stoichiometries(inputs=inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:785 in get_min_max_stoichiometries
    sol = self.solve(all_inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:420 in solve
    energy = self.theoretical_energy_integral(energy_inputs)

  File myPath/lib64/python3.11/site-packages/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py:867 in theoretical_energy_integral
    Vs = self.parameter_values.evaluate(

  File myPath/lib64/python3.11/site-packages/pybamm/parameters/parameter_values.py:869 in evaluate
    raise ValueError(

ValueError: symbol must evaluate to a constant scalar or array
    
```

When I set Entropic change to "0", all works fine. Can you help?
